Changes to uGMT emulator as well as BMTF, uGMT, and uGT unpackers for Kalman muons

dinyar opened the pull request:
This PR contains the following changes:

- The GMTInternalMuon and Muon data format gain fields for dXY ([set]hwDXY()) and unconstrained pT ([set]hwPtUnconstrained())
- The uGMT emulator fills the above fields from the RegionalMuonCand
- The uGMT emulator ghost busting for barrel muons switches to a different algorithm based on the uGMT firmware version. (This accounts for the fact that track finding is different in the Kalman Filter.)
- The unpacker for BMTF outputs/uGMT inputs can unpack unconstrained pT and dXY. It also correctly unpacks the track addresses for the KBMTF. It is backwards compatible by detecting the algo version.
- The unpacker for uGMT/intermediates/uGMT outputs/uGT inputs can unpack unconstrained pT and dXY (and was adapted to the new readout record format). It is also backwards compatible by using the algo version.

It also includes the changes done in PRs #835 (formerly #770) and #768.

@rekovic @thomreis @panoskatsoulis

dinyar commented:
An immediate thing I am not sure about is the input tag chosen in [1] for the BMTF muons. In this PR it is set to simKBmtfDigis, however I'm not sure how best to write this to make the emulator remain backwards compatible.

Any ideas?


[1] https://github.com/cms-l1t-offline/cmssw/pull/836/files?diff=split&w=1#diff-c44432eae1061c9da48110e80b976bbc

panoskatsoulis commented:
Hi Dinyar,
BMTF emulators are 2 modules - one tags its product `simKBmtfDigis` and the other `simBmtfDigis`.

Changing this tag you force the emulation chain to use the Kalman output.
Whether we need to use `simKBmtfDigis` or `simBmtfDigis` depends on what we want the GMT to get as input.

If we still want to emulate according to the old algorithm, then this must not be changed

panoskatsoulis commented:
At some point, we'll need to change this for Run 3.
I suppose that the emulation won't be backwards compatible because GMT will be getting as input statically only Kalman.

**However, there is a solution to this.** 
We had faced this issue in the online DQM, and we solved it.
We can use the same way to fix it.

thomreis commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFUnpackerOutput.cc:
```diff
@@ -13,78 +13,76 @@ namespace l1t
+           RegionalMuonRawDigiTranslator::fillRegionalMuonCand(muCand, raw_first, raw_secnd, processor, tftype::bmtf, isKalman);
+
+           if (muCand.hwQual() == 0)
+           continue;
```
Indentation.

dinyar replied:
Will change it. However now that I look at that line a bit closer: In the GMT we actually use pT to decide if it's an empty muon, in the BMTF it looks like they use quality.. It shouldn't make a difference (especially given that the emulator checks for pT == 0 now), but maybe an opportunity to homogenise?

thomreis replied:
Good point.

dinyar replied:
@panoskatsoulis, can you have a quick look? Can we change this to check for the pT, analogous to the GMT unpacker[1]?

[1] https://github.com/dinyar/cmssw/blob/ugmt_kalman-adaptions/EventFilter/L1TRawToDigi/plugins/implementations_stage2/RegionalMuonGMTUnpacker.cc#L86

panoskatsoulis replied:
Hi @dinyar, I have to check what is the default Pt in case there is no muon 
(in both algos since we don't want to break backwards compatibility)
I think it's possible, I'll come back to you

panoskatsoulis replied:
So, seems that indeed we can homogenize the code. 
Check above my suggestion for these lines.
I checked the config for zero-suppression in the keys and zero-suppresses the record if all 6 words are 0x0.

thomreis commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFUnpackerOutput.cc:
```diff
@@ -13,78 +13,76 @@ namespace l1t
-	  if (muCand.hwQual() == 0)
-	    continue;
-
-	  muCand.setLink(48 + processor);	//the link corresponds to the uGMT input
```
Does the link not need to be set anymore?

dinyar replied:
I noticed that it's only set in the BMTF unpacker (and not the GMT regional muon unpacker) and then saw that it's anyway done indirectly in fillRegionalMuonCand()[1] because setTFIdentifiers() seems to do it[2] (for reference, setLink() is at [3]. Can you have a look to make sure I didn't miss something?

[1] https://github.com/dinyar/cmssw/blob/788557afd732cfc0c4afa1c30d10f6e4c8aff99a/L1Trigger/L1TMuon/src/RegionalMuonRawDigiTranslator.cc#L88
[2] https://github.com/dinyar/cmssw/blob/788557afd732cfc0c4afa1c30d10f6e4c8aff99a/DataFormats/L1TMuon/src/RegionalMuonCand.cc#L18
[3] https://github.com/dinyar/cmssw/blob/788557afd732cfc0c4afa1c30d10f6e4c8aff99a/DataFormats/L1TMuon/interface/RegionalMuonCand.h#L81

thomreis replied:
Looks redundant indeed.

thomreis commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/RegionalMuonGMTUnpacker.h:
```diff
@@ -10,10 +10,13 @@ namespace l1t {
       class RegionalMuonGMTUnpacker : public Unpacker {
          public:
             bool unpack(const Block& block, UnpackerCollections *coll) override;
+            void setKalmanAlgoTrue() { isKalman_ = true; }
```
Is there never going to be a need for a setKalmanAlgoFalse?

dinyar replied:
I initialise the object to isKalman_ = false in [1], similar to the way this is done for the BMTF[2] for consistency. I think once one has made the choice that it should be unpacking Kalman muons one won't need to switch it back off, right?

[1] https://github.com/dinyar/cmssw/blob/788557afd732cfc0c4afa1c30d10f6e4c8aff99a/EventFilter/L1TRawToDigi/plugins/implementations_stage2/RegionalMuonGMTUnpacker.h#L19
[2] https://github.com/dinyar/cmssw/blob/788557afd732cfc0c4afa1c30d10f6e4c8aff99a/EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFUnpackerOutput.h#L14

thomreis replied:
That was my guess as well. As long as all processors use the same FW that should be OK then.

panoskatsoulis commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFUnpackerOutput.cc:
```diff
@@ -13,78 +13,76 @@ namespace l1t
+           RegionalMuonCand muCand;
+           RegionalMuonRawDigiTranslator::fillRegionalMuonCand(muCand, raw_first, raw_secnd, processor, tftype::bmtf, isKalman);
+
+           if (muCand.hwQual() == 0)
```
```suggestion
           if (muCand.hwPt() == 0)
```

dinyar replied:
Thanks for this! You could move a slightly modified check a bit higher (before filling the muon) to avoid working on a known empty muon. (See e.g. https://github.com/dinyar/cmssw/blob/ugmt_kalman-adaptions/EventFilter/L1TRawToDigi/plugins/implementations_stage2/RegionalMuonGMTUnpacker.cc#L86) I don't think it would make a significant difference though, let me know if you'd like to change it or if I should just commit your change as it currently is.

panoskatsoulis replied:
Hi Dinyar, and sorry for that late reply
lets proceed on committing the code as above, and I can modify it later
what you propose is a good change

dinyar replied:
Ok, sounds good!

dinyar commented:
> At some point, we'll need to change this for Run 3.
> I suppose that the emulation won't be backwards compatible because GMT will be getting as input statically only Kalman.
> 
> **However, there is a solution to this.**
> We had faced this issue in the online DQM, and we solved it.
> We can use the same way to fix it.

Hi @panoskatsoulis, that sounds good. Can you point me at how you did this?

panoskatsoulis commented:
yes, there is an extra DQM module, that checks the fwver, and creates 2 collections BMTF1 and BMTF2 
(above specific number means Kalman after an agreement we had before with Stavros)
this is the module
https://github.com/cms-sw/cmssw/tree/master/DQM/L1TMonitor/plugins
You can use exactly the same plugin for BMTF actually, it will make it work
here is the sequence
https://github.com/cms-sw/cmssw/blob/master/DQM/L1TMonitor/python/L1TStage2Emulator_cff.py#L84-L95

dinyar commented:
Ah, thanks for this! So, I think for the emulator this might not be the right approach because e.g. when running on MC we wouldn't have the firmware version.

I can think of two options: Configure the input collection based on the uGMT firmware version number from the conditions data or using a different sequence for the two periods.

What do you think? (Also @thomreis and @rekovic probably have useful input.)

thomreis commented:
I think it would be possible to have such a module that checks the FW version in the GT and then switches the two outputs if needed.
This would also be available when running on MC. Of course it would have to be thought of beforehand what to version to put in the GT. I do not see a simple way to override this afterwards if one wants to rerun with a different algo on a given MC sample.

abrinke1 commented:
Hi Vladimir, all,

Perhaps the global approach to Era and/or GT-based configurations for
"future" L1T configurations in development would be an appropriate topic
for the L1T Offline Software meeting.  We recently ran into a similar issue
in EMTF:
https://github.com/cms-sw/cmssw/pull/29289

And the CSC group is currently updating CSC LCT and GEM formats for Run 3:
https://github.com/cms-l1t-offline/cmssw/pull/818

We want to make sure these are all in synch, and part of that means having
the same switches control the new TP input formats as well as L1T emulator
behavior.

Best regards,
Andrew


On Tue, Mar 24, 2020 at 10:56 AM Thomas Reis <notifications@github.com>
wrote:

> I think it would be possible to have such a module that checks the FW
> version in the GT and then switches the two outputs if needed.
> This would also be available when running on MC. Of course it would have
> to be thought of beforehand what to version to put in the GT. I do not see
> a simple way to override this afterwards if one wants to rerun with a
> different algo on a given MC sample.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-l1t-offline/cmssw/pull/836#issuecomment-603324707>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHVOUD2MZZR2NTBIBRRO6DRJDJUXANCNFSM4LCNR4AA>
> .
>

abrinke1 commented:
Correction: the LCT data format update I was thinking of was this:
https://github.com/cms-sw/cmssw/pull/28846

Which Sven is working on expanding here:
https://github.com/cms-sw/cmssw/compare/CMSSW_11_1_X...dildick:CMSSW_11_1_X_2020-03-10-1100-extend-clct-definition?expand=1

On Thu, Mar 26, 2020 at 11:09 AM Andrew Brinkerhoff <andrewbrink@gmail.com>
wrote:

> Hi Vladimir, all,
>
> Perhaps the global approach to Era and/or GT-based configurations for
> "future" L1T configurations in development would be an appropriate topic
> for the L1T Offline Software meeting.  We recently ran into a similar issue
> in EMTF:
> https://github.com/cms-sw/cmssw/pull/29289
>
> And the CSC group is currently updating CSC LCT and GEM formats for Run 3:
> https://github.com/cms-l1t-offline/cmssw/pull/818
>
> We want to make sure these are all in synch, and part of that means having
> the same switches control the new TP input formats as well as L1T emulator
> behavior.
>
> Best regards,
> Andrew
>
>
> On Tue, Mar 24, 2020 at 10:56 AM Thomas Reis <notifications@github.com>
> wrote:
>
>> I think it would be possible to have such a module that checks the FW
>> version in the GT and then switches the two outputs if needed.
>> This would also be available when running on MC. Of course it would have
>> to be thought of beforehand what to version to put in the GT. I do not see
>> a simple way to override this afterwards if one wants to rerun with a
>> different algo on a given MC sample.
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/cms-l1t-offline/cmssw/pull/836#issuecomment-603324707>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABHVOUD2MZZR2NTBIBRRO6DRJDJUXANCNFSM4LCNR4AA>
>> .
>>
>

rekovic commented:
hi @dinyar, all, 

We designed the BMTF unpacker to use the fw version available via GlobalTag.  And that is fine since the RAW data from all the Runs taken have this useful  information stored in the DB. 

However, what we do with DIGI data, which become inputs to a subsystem(s) of L1T chain, is a different story.  How do we use the available DIGI data, unpacked or simulated, in what configuration of the L1T trigger system or parts thereof do we want to propagate them downstream, we control by the Era modifiers.  The era modifiers we introduced are used to configure L1T emulators to describe a hardware design of our L1T system. 

We use Era modifiers available :
- `stage2L1Trigger` to distinguish Stage2 from Legacy , 
- `stage2L1Trigger_2017` to distinguish 2017 from 2016

The ability to emulation different scenarios needs to preserved. Please roll-back the line [1]. 

Also make sure that the following workflow with unpacker and emulator is not broken 
```
cmsDriver.py --python_filename=l1t_test.py -s RAW2DIGI --era=Run2_2016 --customise=L1Trigger/Configuration/customiseReEmul.L1TReEmulFromRAW  --customise=L1Trigger/Configuration/customiseUtils.L1TTurnOffUnpackStage2GtGmtAndCalo --conditions=auto:run2_data -n 10 --data  --no_output --filein=/store/data/Run2016H/HLTPhysics3/RAW/v1/000/281/602/00000/0E71643D-BF82-E611-8593-02163E011B19.root
```


When it comes to making emulator configuration change [1] needed for future-studies of Run3, this will be done using Era modifier in the simulation sequence. 

[1] https://github.com/cms-l1t-offline/cmssw/pull/836/files?diff=split&w=1#diff-c44432eae1061c9da48110e80b976bbc

dinyar commented:
Hi @rekovic,

Just for my understanding, is the following what you propose?

- Unpackers continue to be configured by the FW version in the readout record
- Packers are configured by era (This is my extrapolation.)
- Emulators are configured by era (The uGMT emulator is currently (largely?) configured by the firmware version from the event setup and I understand this can either be set by the Global Tag or via the eras?)

I think a few other systems had similar issues, would it maybe be useful to discuss this in a offline software meeting as @abrinke1 suggested?

> The ability to emulation different scenarios needs to preserved. Please roll-back the line [1].
> 
> Also make sure that the following workflow with unpacker and emulator is not broken
> 
> ```
> cmsDriver.py --python_filename=l1t_test.py -s RAW2DIGI --era=Run2_2016 --customise=L1Trigger/Configuration/customiseReEmul.L1TReEmulFromRAW  --customise=L1Trigger/Configuration/customiseUtils.L1TTurnOffUnpackStage2GtGmtAndCalo --conditions=auto:run2_data -n 10 --data  --no_output --filein=/store/data/Run2016H/HLTPhysics3/RAW/v1/000/281/602/00000/0E71643D-BF82-E611-8593-02163E011B19.root
> ```
> 
> When it comes to making emulator configuration change [1] needed for future-studies of Run3, this will be done using Era modifier in the simulation sequence.
> 
> [1] https://github.com/cms-l1t-offline/cmssw/pull/836/files?diff=split&w=1#diff-c44432eae1061c9da48110e80b976bbc

Sure, will do and let you know!

Cheers,
Dinyar

rekovic commented:
Thanks @dinyar .

On purpose I have not mentioned the packers here, since this PR does not contain it and I did not want to overload you here in this PR.   I presume the packer will come in some of next following PRs. The packers need to be in synch with the design of the hardware system that the L1T software emulates.  Therefore, the packers must be Era configurable.   

Yes, sure, we will discuss the topic of Eras in the next offline SW meeting, which will be next Wednesday.

dinyar commented:
Hi @rekovic,

I reverted the input collections and will make this configurable with eras similar to your example from the meeting today. 

Regarding the workflow you asked me to check: As briefly hinted on Skype it looks to me like this workflow never calls the uGMT emulator -- I tried to confirm this by modifying the emulator and saw no changes in the output products. It's a bit puzzling to me -- when I change the unpacker I see changes in the output EDM file, but when I change the emulator I don't. From the customisation commands I was expecting the opposite behaviour?

Checking the workflow anyway uncovered a bug in the unpacker I had introduced which prompted me to rewrite for clarity (@thomreis, could you have a look again?).

dinyar commented:
Hi @rekovic,

I now tested the workflow you indicated via the L1Ntuple[1] as well as those for 2017[2] and 2018[3] data, comparing the results with 10_6_0 without my changes and things look the same.

Two things struck me that are almost certainly misconceptions on my side as I also see them in the unchanged 10_6_0:
- For the 2016 workflow it seems that re-emulated EMTF muons never reach the GMT. I see them in the l1UpgradeTfMuonEmuTree, but I then don't see any EMTF muons in the l1UpgradeEmuTree. Is this expected?
- When I run edmDumpEventContent on the EDM file I don't see sim*Digis, but instead e.g. gmtStage2Digis. I thought "L1TTurnOffUnpackStage2GtGmtAndCalo" would mean that the opposite should happen?

Finally two questions on how to proceed:
- I'd like to test this PR with displaced information from the BMTF. Is there a recipe to run the KBMTF emulator already? (Maybe @panoskatsoulis can also help here?)
- Would you like me to add the configuration of the emulator (and also the packer?) to this PR? Now that we have a plan it might be more logical to have all these things in one place.

Cheers,
Dinyar

[1] cmsDriver.py --python_filename=l1t_test.py -s RAW2DIGI --era=Run2_2016 --customise=L1Trigger/Configuration/customiseReEmul.L1TReEmulFromRAW  --customise=L1Trigger/Configuration/customiseUtils.L1TTurnOffUnpackStage2GtGmtAndCalo --customise=L1Trigger/L1TNtuples/customiseL1Ntuple.L1NtupleRAWEMU  --conditions=auto:run2_data -n 100 --data --filein=/store/data/Run2016H/HLTPhysics3/RAW/v1/000/281/602/00000/0E71643D-BF82-E611-8593-02163E011B19.root
[2] cmsDriver.py l1Ntuple -s RAW2DIGI --python_filename=data.py -n 420 --no_output --era=Run2_2017 --data --conditions=101X_dataRun2_Prompt_v9 --customise=L1Trigger/Configuration/customiseReEmul.L1TReEmulFromRAW --customise=L1Trigger/L1TNtuples/customiseL1Ntuple.L1NtupleRAWEMU --customise=L1Trigger/Configuration/customiseSettings.L1TSettingsToCaloStage2Params_2017_v1_8_4 --filein=file:/eos/cms/store/data/Run2017F/SingleMuon/RAW/v1/000/306/125/00000/FE50A2E4-B1C0-E711-BFA3-02163E014781.root
[3] cmsDriver.py l1Ntuple -s RAW2DIGI --python_filename=data.py -n 240 --no_output --era=Run2_2018 --data --conditions=103X_dataRun2_Prompt_v3 --customise=L1Trigger/Configuration/customiseReEmul.L1TReEmulFromRAW --customise=L1Trigger/L1TNtuples/customiseL1Ntuple.L1NtupleAODRAWEMU --customise=L1Trigger/Configuration/customiseSettings.L1TSettingsToCaloParams_2018_v1_4 --filein=file:/eos/cms/store/data/Run2018D/SingleMuon/RAW-RECO/ZMu-PromptReco-v2/000/321/457/00000/60D23B7B-C7A5-E811-B14B-FA163EF4F4A1.root

eyigitba commented:
Hi @dinyar ,

>  For the 2016 workflow it seems that re-emulated EMTF muons never reach the GMT. I see them in the l1UpgradeTfMuonEmuTree, but I then don't see any EMTF muons in the l1UpgradeEmuTree. Is this expected?

Does the unchanged 10_6_0 branch you use include the recent fixes to EMTF emulation? If not maybe this is due to the problem we discovered while validating UL16 MC.

Best,
Efe

panoskatsoulis commented:
Hi @dinyar 
for the BMTF side:
- No there isn't any general L1 machinery ready to run the Kalman emulator out of the box. In general, the "recipe" is: replace "simBmtfDigis" with "simKBmtfStubs*simKBmtfDigis" in your cmssw configuration files
- Let's keep them separated, I prefer it. There is another guy who started playing with the BMTF software too (esp Packers/Unpackers), so keeping it separated will make it easier for him to follow.
Is it fine for you to open a new PR about this? (@rekovic is this fine with you too?)

dinyar commented:
> Hi @dinyar ,
> 
> > For the 2016 workflow it seems that re-emulated EMTF muons never reach the GMT. I see them in the l1UpgradeTfMuonEmuTree, but I then don't see any EMTF muons in the l1UpgradeEmuTree. Is this expected?
> 
> Does the unchanged 10_6_0 branch you use include the recent fixes to EMTF emulation? If not maybe this is due to the problem we discovered while validating UL16 MC.
> 
> Best,
> Efe

Ah, that may be the case. I'll try again with l1t-integration-v103.2, this contains the fixes already, right?

Cheers,
Dinyar

dinyar commented:
Hi @panoskatsoulis 

> * No there isn't any general L1 machinery ready to run the Kalman emulator out of the box. In general, the "recipe" is: replace "simBmtfDigis" with "simKBmtfStubs*simKBmtfDigis" in your cmssw configuration files

That worked, thanks! 

> * Let's keep them separated, I prefer it. There is another guy who started playing with the BMTF software too (esp Packers/Unpackers), so keeping it separated will make it easier for him to follow.
>   Is it fine for you to open a new PR about this? (@rekovic is this fine with you too?)

Sure, no problem for the packer from my side. For the emulator I'll give it a quick think, but I think it shouldn't be a problem either. Anyway, lets see if @rekovic has a preference.

Cheers,
Dinyar

eyigitba commented:
> > Hi @dinyar ,
> > > For the 2016 workflow it seems that re-emulated EMTF muons never reach the GMT. I see them in the l1UpgradeTfMuonEmuTree, but I then don't see any EMTF muons in the l1UpgradeEmuTree. Is this expected?
> > 
> > 
> > Does the unchanged 10_6_0 branch you use include the recent fixes to EMTF emulation? If not maybe this is due to the problem we discovered while validating UL16 MC.
> > Best,
> > Efe
> 
> Ah, that may be the case. I'll try again with l1t-integration-v103.2, this contains the fixes already, right?
> 
> Cheers,
> Dinyar

Yes  l1t-integration-v103.2 should contain the fix.

dinyar commented:
Superseded by #838.
